src/QsCompiler/Transformations/CallGraphWalker.cs

```diff
@@ -17,6 +16,256 @@ namespace Microsoft.Quantum.QsCompiler.Transformations.CallGraphWalker
+                {
+                    var scc = new HashSet<int>();
+
+                    var minNode = node;
+                    int nodeInScc;
+                    do
+                    {
+                        nodeInScc = nodeStack.Pop();
+                        var wInfo = nodeInfo[nodeInScc];
+                        wInfo.OnStack = false;
+                        nodeInfo[nodeInScc] = wInfo;
+                        scc.Add(nodeInScc);
+
+                        // Keep track of minimum node in scc
+                        if (minNode > nodeInScc)
+                        {
+                            minNode = nodeInScc;
+                        }
+
+                    } while (node != nodeInScc);
+                    SCCs.Add((scc, minNode));
+                }
+            }
+
+            foreach (var node in inputGraph.Keys)
+            {
+                if (!nodeInfo.ContainsKey(node))
+                {
+                    strongconnect(node);
+                }
+            }
+
+            return SCCs;
+        }
+
+        /// <summary>
+        /// Johnson's algorithm for finding all cycles in an SCC, or strongly-connected component.
+        ///
+        /// It will process an individual SCC for cycles by looking at each of their nodes
+        /// starting with the minimum-valued node of the SCC as the current node. Children of
+        /// the current node are explored, and if any of them are the starting node, a cycle is
+        /// found. Nodes are marked as 'blocked' when visited, and if no cycle is found after
+        /// exploring a node's children, the node is marked as being blocked on its children.
+        /// Only when a cycle is found is a node unblocked and any node blocked on that node is
+        /// also unblocked, recursively.
+        /// </summary>
+        private List<List<int>> GetSccCycles(Dictionary<int, List<int>> intputSCC, int startNode)
+        {
+            var cycles = new List<List<int>>();
+            var blockedSet = new HashSet<int>();
+            var blockedMap = new Dictionary<int, HashSet<int>>();
+            var nodeStack = new Stack<int>();
+
+            void unblock(int node)
+            {
+                if (blockedSet.Remove(node) && blockedMap.TryGetValue(node, out var nodesToUnblock))
+                {
+                    blockedMap.Remove(node);
+                    foreach (var n in nodesToUnblock)
+                    {
+                        unblock(n);
+                    }
+                }
+            }
+
+            bool populateCycles(int currNode)
+            {
+                var foundCycle = false;
+                nodeStack.Push(currNode);
+                blockedSet.Add(currNode);
+
+                foreach (var child in intputSCC[currNode])
+                {
+                    if (child == startNode)
+                    {
+                        foundCycle = true;
+                        cycles.Add(nodeStack.Reverse().ToList());
```

Comment:
Why is it necessary to reverse the stack?

Reply:
The cycles will be backwards if I don't. I'm not sure if there was something else to be done here.